content/client-lib-development-guide/features.textile

```diff
@@ -222,6 +222,10 @@ h3(#rest-channel). Channel
 *** @(RSL1g4)@ When publishing a message with an explicit @clientId@ that is incompatible with the identified client's @clientId@ (either inferred or explicitly configured in the token or @ClientOptions@), the library will reject the message immediately and indicate an error
 ** @(RSL1h)@ Where the library language permits, the @Channel#publish(name, data)@ method should provide an optional argument that allows the @clientId@ value to be specified such as @Channel#publish('event', 'data', { clientId: 'John' })@, and/or the @extras@ field to be specified such as @Channel#publish('event', 'data', { extras: { push: { 'key': 'value' } })@
 ** @(RSL1i)@ If the total size of the message or (if publishing an array) messages, calculated per "TO3l8":#TO3l8, exceeds the @maxMessageSize@, then the client library should reject the publish and indicate an error
+** @(RSL1j)@ When @Message@ objects are provided, any valid @Message@ attribute (that is, an attribute specified in "TM2":#TM2) that is supplied by the caller must be included in the encoded message. (This does not mean it must be included _unaltered_; for example the @data@ and @encoding@ will be subject to processing per "RSL4":#RSL4)
+*** @(RSL1j1)@ A test should exist which publishes a @Message@ which includes a fixed @id@ twice, and checks that it is received only once
+** @(RSL1k)@ If @idempotentRestPublishing@ (see "TO3n":#TO3n) is enabled, if the @Message@s do not include caller-supplied @ids@, for each @Message@ in the array the library should randomly generate get 9 bytes from a source of randomness, encode them as base64, and set it as the @Message@ @id@
```

Comment:
Odd that you chose 9 bytes, any thinking?

Reply:
We are going to go with instead that a random ID is generated, and every message is given the ID + `:` + index within the array of messages being published.

Reply:
We need to specify that a test should exist that publishes an array of messages to an endpoint that fails the first time, and succeeds upon the second request.  Each attempt must contain the exact same payload with the same generated IDs.

Reply:
>Odd that you chose 9 bytes, any thinking?

base64 is 6 bits per character, so number of bits needs to be divisible by 6 to have it be 'clean', else you get variable numbers of `=`s at the end for padding. So 72 bits (9 bytes; 12 characters) works, as does 96 (12 bytes; 16 characters); Paddy commented that 16 characters was a bit long, so 9 bytes it was.

Reply:
Ok, this makes perfect sense, :+1: